### PR TITLE
workload: fix rand workload

### PR DESCRIPTION
**Pull request by renatolabs**
The `rand` workload generates random table definitions when
intialized, and performs random writes to the first table created when
run. It relies on the `randgen` package to generate random data and
has internal logic to convert from Datum to the corresponding Go data
structure to be passed to a database call.

This means that whenever `randgen` is able to generate new types of
data, the `rand` workload needs to be changed accordingly. Since there
were no automated means to detect when the `rand` workload had
drifted (and this workload is probably not run very frequently),
running `rand` could fail because it wouldn't know how to generate a
Go data structure for a specific datum.

This commit updates the `rand` workload to add support for the missing
data types and also adds a unit test that verifies that it is able to
generate and insert data for all possible column types. If a new type
is added to `randgen.RandomDatum`, this test should fail.

This also changes the random seed used in `rand` in each run.

### CODE REVIEW

**Comment by renatolabs** [comment]
Note that this does not guarantee that no errors are seen when running the `rand` workload. One issue I observed in my tests is that random data generation is not constrained by the schema definition. In other words, the following is possible:

* table is created with column `c1` as `INT8`, and column `c2` is defined as `c1 + 200`
* `randgen` generates `MaxInt64-1` as a value for `c1`
* error: `c2` overflows

Fixing this type of "bug" is out of scope for this PR as it's not related to the workload itself.

File: pkg/workload/rand/rand.go [hunk 1]
```diff
@@ -67,10 +85,11 @@ var randMeta = workload.Meta{
 		g.flags.StringVar(&g.tableName, `table`, ``, `Table to write to`)
 		g.flags.IntVar(&g.batchSize, `batch`, 1, `Number of rows to insert in a single SQL statement`)
 		g.flags.StringVar(&g.method, `method`, `upsert`, `Choice of DML name: insert, upsert, ioc-update (insert on conflict update), ioc-nothing (insert on conflict no nothing)`)
-		g.flags.Int64Var(&g.seed, `seed`, 1, `Key hash seed.`)
+		g.flags.Int64Var(&g.seed, `seed`, defaultRandomSeed(), `Key hash seed.`)
```
Comment (srosenberg) [inline pkg/workload/rand/rand.go:88]: Nice! I wasn't aware we were defaulting to `1`, so I checked other workloads. They seem to follow the same pattern, e.g., https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/tpcc/tpcc.go#L187

I wonder why?

File: pkg/workload/rand/BUILD.bazel [hunk 2]
```diff
@@ -21,4 +23,24 @@ go_library(
     ],
 )
 
+go_test(
+    name = "rand_test",
+    srcs = ["rand_test.go"],
+    args = ["-test.timeout=295s"],
```
Comment (srosenberg) [inline pkg/workload/rand/BUILD.bazel:29]: Bazel defaults to "medium" but it seems like this test doesn't need more than a few seconds, so maybe `size = "small"?

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

___
*[`pkg/workload/rand/BUILD.bazel` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDu-r2-29:-NCacfJw85Kjd8GWN129:b-l97i80) ([raw file](https://github.com/cockroachdb/cockroach/blob/209cf14ebaee40e0bfab7d37c699b5c3e0bd18c2/pkg/workload/rand/BUILD.bazel#L29)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Bazel defaults to "medium" but it seems like this test doesn't need more than a few seconds, so maybe `size = "small"`?
</blockquote></details>

Good point, fixed.
___
*[`pkg/workload/rand/rand.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDv-r2-88:-NCactXg41fHGfLGV865:b-lx9w5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/209cf14ebaee40e0bfab7d37c699b5c3e0bd18c2/pkg/workload/rand/rand.go#L88)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Nice! I wasn't aware we were defaulting to `1`, so I checked other workloads. They seem to follow the same pattern, e.g., https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/tpcc/tpcc.go#L187

I wonder why?
</blockquote></details>

Yes, after making this change I realized the other workloads also use `1` as the default seed. IMO, making the seed different on each run is a better default though. Especially in the context of workloads being used in tests.

**Review by smg260 — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

**Review by srosenberg — COMMENTED** [review]
Indeed, it doesn't attempt to satisfy types or constraints; adding a couple more examples from my local runs,

```
Error: pq: duplicate key value violates unique constraint "table6234468500453514806_col6234468500453514806_4_expr_key"
Error: pq: bit string length 1 does not match type BIT(36)
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-:-NCc4bCD3htLgNJ1S4MB:b7unu6l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/workload/rand/rand.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDv-r2-88:-NCc5uqxAx57gq4tZ5kS:bwjm8nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/209cf14ebaee40e0bfab7d37c699b5c3e0bd18c2/pkg/workload/rand/rand.go#L88)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Yes, after making this change I realized the other workloads also use `1` as the default seed. IMO, making the seed different on each run is a better default though. Especially in the context of workloads being used in tests.
</blockquote></details>

I agree that it shouldn't default to a fixed value. I am just surprised it's been like this since this commit [1] (circa 2019). We could have been fuzzing more states :(
We should update all workloads to use a new seed by default; obviously, it's out of scope for this PR.

[1] https://github.com/cockroachdb/cockroach/commit/626b1a6816df14fc60e273f1a93264c162e341b8

**Review by srosenberg — COMMENTED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/workload/rand/rand.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCc9V4v5xnv3B1AmO3V:-NCc9V4v5xnv3B1AmO3W:b-paneav) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec6b1acac5d1f395ff371ba882ac6a7ead288ab/pkg/workload/rand/rand.go#L67)):*
> ```Go
> func defaultRandomSeed() int64 {
> 	defaultSeedOnce.Do(func() {
> 		defaultSeed = timeutil.Now().UTC().UnixNano()
> ```

nit: `UTC` is redundant.

On second thought, a time-based seed is not as uniform as a crypto-based seed. `cli.Main` is already using a seed by reading it from `/dev/urandom` [1]. I think we should do the same in all the workloads.

[1] https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/cli.go#L53 
___
*[`pkg/workload/rand/rand_test.go` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCcB69JEdHCp2nJCsaH:-NCcB69JEdHCp2nJCsaI:b-z52269) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec6b1acac5d1f395ff371ba882ac6a7ead288ab/pkg/workload/rand/rand_test.go#L44)):*
> ```Go
> // insert it into a column
> func TestRandRun(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

Let's also add `defer log.Scope(t).Close(t)` since there are quite a few subtests; collectively, their output is verbose. In case of failure, it's nice to have a dedicated log with all the output.

**Review by srosenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/workload/rand/rand.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDv-r2-88:-NCcDLCS4981IPLZqPFE:b-qvlumi) ([raw file](https://github.com/cockroachdb/cockroach/blob/209cf14ebaee40e0bfab7d37c699b5c3e0bd18c2/pkg/workload/rand/rand.go#L88)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I agree that it shouldn't default to a fixed value. I am just surprised it's been like this since this commit [1] (circa 2019). We could have been fuzzing more states :(
We should update all workloads to use a new seed by default; obviously, it's out of scope for this PR.

[1] https://github.com/cockroachdb/cockroach/commit/626b1a6816df14fc60e273f1a93264c162e341b8
</blockquote></details>

Note, another side-effect of this change is that the workload binary needs to be invoked with `--init`. Previously, it was possible to reuse the same tables between runs. I don't think we need to preserve that behavior, just noting the (obvious) difference :)

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)

___
*[`pkg/workload/rand/rand.go` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDv-r2-88:-NCefp99BcOXFhQDURW2:bbmv1c0) ([raw file](https://github.com/cockroachdb/cockroach/blob/209cf14ebaee40e0bfab7d37c699b5c3e0bd18c2/pkg/workload/rand/rand.go#L88)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Note, another side-effect of this change is that the workload binary needs to be invoked with `--init`. Previously, it was possible to reuse the same tables between runs. I don't think we need to preserve that behavior, just noting the (obvious) difference :)
</blockquote></details>

It doesn't need to be run with `--init`; but the caller does need to make sure they pass the same seed that was used when initializing. If a user is running the commands interactively, they can copy and paste the seed that is logged. If calling `workload` programatically, the best way to go is to probably generate the seed out of band and pass it to every `workload` call.
___
*[`pkg/workload/rand/rand.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCc9V4v5xnv3B1AmO3V:-NCefjjE76NNv7GslNrS:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec6b1acac5d1f395ff371ba882ac6a7ead288ab/pkg/workload/rand/rand.go#L67)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

nit: `UTC` is redundant.

On second thought, a time-based seed is not as uniform as a crypto-based seed. `cli.Main` is already using a seed by reading it from `/dev/urandom` [1]. I think we should do the same in all the workloads.

[1] https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/cli.go#L53 
</blockquote></details>

Good catch, fixed.
___
*[`pkg/workload/rand/rand_test.go` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCcB69JEdHCp2nJCsaH:-NCegPtD5UUCj4macgE7:b-npjcwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec6b1acac5d1f395ff371ba882ac6a7ead288ab/pkg/workload/rand/rand_test.go#L44)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Let's also add `defer log.Scope(t).Close(t)` since there are quite a few subtests; collectively, their output is verbose. In case of failure, it's nice to have a dedicated log with all the output.
</blockquote></details>

Good point, added.

**Review by renatolabs — COMMENTED** [review]
> Reviewed 1 of 3 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCelbshCndxwXZ02Mf9:-NCelbshCndxwXZ02MfA:bw1g1ui) (no related file):*
Created issue https://github.com/cockroachdb/cockroach/issues/88566 to deal with the default seeds in workloads.

___

**Review by renatolabs — COMMENTED** [review]
Yes, the first one is a conflict on a non-primary-key index. It's technically possible to deal with them (the workload already deals with conflicts on the primary key), but it doesn't do that right now.

As for the second one, it's not that it doesn't try to satisfy the type, but more that the type OID read from `pg_catalog.pg_attribute` for the `BIT` type maps to a variable-length `BIT` type even though the schema was defined with as `BIT` of a specific length. I'm not entirely sure why that is the case.

@ajwerner would you know if there's a reliable way for this workload to get the width of the `BIT` column? I'm asking you directly because I copied a change you made in the `schemachange` workload [1] in this PR, and I don't understand why it's needed.

[1] https://github.com/cockroachdb/cockroach/blob/975701e0fa4c3a3c8b78e230ad5abd1161cf215f/pkg/workload/schemachange/type_resolver.go#L106-L111

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-:-NCeyhJxE6ZgREgqcxE_:b-mb9dng)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @srosenberg)

**Comment by ajwerner** [comment]
The OIDs don't imply the widths. I guess you can make a char(2) type or whatever width you want but if you don't name the width I guess we do 1? I think I just wanted to make sure randgen always made a valid value and 1 will do that for all char widths. 

I've learned now that you can access column widths via the `information_schema.columns` table's `character_maximum_length` column. That has to be done on a per column basis.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCff5KG9fNbEOf5Yeoh:-NCff5KG9fNbEOf5Yeoi:b-mc4f75) (no related file):*
Pushed a couple of changes here:

* Improved logging of seed, especially in the case where the caller passes `--seed`. Not great, but current solution works. We should come up with a better, reusable strategy in #88566.
* Changed the test to actually test all possible column types that `randgen` can generate; specifically, arrays of different types of data are now part of the test as well.
* A consequence of the above is that I had to change the logic for arrays; arrays of geography/geometry data types use `:` as separator.
* I used the information on @ajwerner's comment about `information_schema.columns` to get the width of bit/char columns. We shouldn't see mismatched width errors when trying to insert anymore.

@srosenberg or @smg260 could you take another look, please?

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCfhyT_60lknwybRmKG:-NCfhyTa91qod5-Ca3-e:bhdgcln) (no related file):*
@srosenberg taking a second look at the errors you posted:

```
Error: pq: duplicate key value violates unique constraint "table6234468500453514806_col6234468500453514806_4_expr_key"
```

I've seen this happen when I run the workload without first clearing the data from a previous run. As I said, it's technically possible to see this error in other circumstances, but the random generator would have to generate the same value twice, which is quite unlikely for most column types. Could you try running `init` with `--drop` and let me know if the issue persists (if you happen to have access to the seed that was used in this run)?

```
Error: pq: bit string length 1 does not match type BIT(36)
```

This has to do with the wrong width information being used in the `rand` workload, which should now be fixed. If you still have the seed for this one, I'd appreciate if you could double check that the fix works.

___

File: pkg/workload/rand/rand.go [hunk 3]
```diff
@@ -56,6 +64,17 @@ func init() {
 	workload.Register(randMeta)
 }
 
+// defaultRandomSeed sets the `defaultSeed` package variable if it's
+// not yet set. This makes it possible for every run of the `rand`
+// workload to use a different seed by default
+func defaultRandomSeed() int64 {
```
Comment (smg260) [inline pkg/workload/rand/rand.go:70]: Clarification - why is `sync.Once` required over simply binding `randutil.NewPseudoSeed()` to `&g.seed`on L92

**Review by smg260 — APPROVED** [review]
LGTM

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/workload/rand/rand.go` line 70 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/88362#-NCVtj7QXhjwHRkMueDv-r8-70:-ND4hJhJ6Nm7hwJi1SRs:b-rupg97) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f7d54dcd11791e16354f9d0342d7bd807df4ab/pkg/workload/rand/rand.go#L70)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

Clarification - why is `sync.Once` required over simply binding `randutil.NewPseudoSeed()` to `&g.seed`on L92
</blockquote></details>

Because `New` is called multiple times, and I just want to set the default seed to a random value once.

**Comment by renatolabs** [comment]
bors r=smg260

TFTR!

**Pull request merged**